### PR TITLE
frontend: implement UI user notifications

### PR DESCRIPTION
**Pull request by asdine**
This implements the UI changes requested on [Figma](https://www.figma.com/file/0567AnXEpQTzQXkgs7AzEh/%2316510-%E2%80%93-WIP-%E2%80%93-Status-notification-system?node-id=466%3A279).

@quinnkeast I have hardcoded the notifications so that you can test locally without having to trigger them manually, please take a look 👍🏼

### CODE REVIEW

**Comment by quinnkeast** [comment]
(Brilliant idea hard-coding it for local review!)

Some tweaks:
- [ ] Cloning + Available for search + Could not be synced → Although slightly misleading in the copy, the links to "Manage repositories" should instead link to the main Repositories view.
- [ ] While I like the intent, the "success" and "warning" icons should remain as in the design for now, rather than use the toolbar's cloud icon (this might be an artifact of the redesign). That'd mean, as an example,

CleanShot 2021-06-10 at 08 41 35@2x

instead of 

CleanShot 2021-06-10 at 08 41 54@2x

Does this PR account for the "placeholder" states, or will that be a separate implementation?

CleanShot 2021-06-10 at 08 42 56@2x

**Comment by artemruts** [comment]
> (Brilliant idea hard-coding it for local review!)
> 
> Some tweaks:
> 
> * [ ]  Cloning + Available for search + Could not be synced → Although slightly misleading in the copy, the links to "Manage repositories" should instead link to the main Repositories view.
> * [ ]  While I like the intent, the "success" and "warning" icons should remain as in the design for now, rather than use the toolbar's cloud icon (this might be an artifact of the redesign). That'd mean, as an example,
> 
> CleanShot 2021-06-10 at 08 41 35@2x
> 
> instead of
> 
> CleanShot 2021-06-10 at 08 41 54@2x
> 
> Does this PR account for the "placeholder" states, or will that be a separate implementation?
> 
> CleanShot 2021-06-10 at 08 42 56@2x

Hey @quinnkeast Thanks for your feedback!

I'll address the issues today since Asdine is out.
As for placeholder states - I confirm they exists, just wasn't a part of this hardcoded demo.

**Comment by artemruts** [comment]
@tjkandala can you please take a look at [this diff](https://github.com/sourcegraph/sourcegraph/pull/21916/files?file-filters%5B%5D=.go&file-filters%5B%5D=.graphql&file-filters%5B%5D=.snap&file-filters%5B%5D=.ts&file-filters%5B%5D=.tsx#diff-1ef7c0063d5bdfe023abdc4efa1f93eace2421bd5a1293211c9f0fe58bc49383), that's the 99% of the changes for this PR

@quinnkeast do you mind taking another look? There's no hardcoded UI, so you may not reproduce some notifications.

**Review by unknwon — APPROVED** [review]
LGTM for the backend portion.

**Comment by quinnkeast** [comment]
> @quinnkeast do you mind taking another look? There's no hardcoded UI, so you may not reproduce some notifications.

For sure!

Some remarks:

- Should I be reviewing link behaviour right now? If so: "Manage repositories" takes me to the instance repositories section. Does this instead take the user to their own repositories page in the settings when they're on Cloud?
- After removing all code host connections and repositories, it continues to show me the "Repositories up to date" state, even after a refresh. I'd have expected the "empty" states to show—first the connect code host state, then the add repositories state. Is this expected/related to local development?

Visually, looks rad.

File: client/web/src/nav/StatusMessagesNavItem.tsx [hunk 1]
```diff
@@ -159,89 +227,223 @@ export class StatusMessagesNavItem extends React.PureComponent<Props, State> {
...
+            }
+        })
     }
 
     private renderIcon(): JSX.Element | null {
-        const { CloudAlertIcon, CloudSyncIcon, CloudCheckIcon } = iconsToShow(this.props.isRedesignEnabled)
+        const { CloudAlertIcon, SyncIcon, CloudCheckIcon } = iconsToShow(true)
 
         if (isErrorLike(this.state.messagesOrError)) {
             return <CloudAlertIcon className="icon-inline-md" />
```
Comment (tjkandala) [inline client/web/src/nav/StatusMessagesNavItem.tsx:416]: So on network error, we show an alert icon with no tooltip? I wonder if that's confusing

**Review by tjkandala — APPROVED** [review]
Frontend LGTM, some minor nits but seems good to merge!

File: client/web/src/nav/StatusMessagesNavItem.tsx (hunk 1 above)
Comment (artemruts) [reply client/web/src/nav/StatusMessagesNavItem.tsx:416]: That's a great point!

File: client/web/src/nav/StatusMessagesNavItem.tsx [hunk 2]
```diff
@@ -159,89 +227,223 @@ export class StatusMessagesNavItem extends React.PureComponent<Props, State> {
...
+                    linkTo={links.manageCodeHosts}
+                    linkText="Connect with code host"
+                    linkOnClick={this.toggleIsOpen}
+                    entryType="not-active"
+                    isRedesignEnabled={this.props.isRedesignEnabled}
+                />
+            )
+        }
+
+        const cloningProgress = message.find(message_ => message_.type === 'CloningProgress')
```
[the remark this answers is not in the document]
Comment (artemruts) [reply client/web/src/nav/StatusMessagesNavItem.tsx:333]: Possibly found an OK name! :)

**Comment by artemruts** [comment]
> > @quinnkeast do you mind taking another look? There's no hardcoded UI, so you may not reproduce some notifications.
> 
> For sure!
> 
> Some remarks:
> 
> * Should I be reviewing link behaviour right now? If so: "Manage repositories" takes me to the instance repositories section. Does this instead take the user to their own repositories page in the settings when they're on Cloud?
> * After removing all code host connections and repositories, it continues to show me the "Repositories up to date" state, even after a refresh. I'd have expected the "empty" states to show—first the connect code host state, then the add repositories state. Is this expected/related to local development?
> 
> Visually, looks rad.

Hey, Quinn!

I believe the behavior you're describing is consistent with what Admins see on `.com` now (links and no "no code hosts/ no repos" notification), we will address this down the line. You'd have to create a non admin user to fully test the notifications, sorry.

We kept admins with the similar-ish UI because notification changes for Admins (as I understand it) will also propagate to on prem instances.

**Comment by quinnkeast** [comment]
>I believe the behavior you're describing is consistent with what Admins see on .com now (links and no "no code hosts/ no repos" notification), we will address this down the line. You'd have to create a non admin user to fully test the notifications, sorry.

Ahh, perfect, that explains it. We should make note of any specific things like this to help with reviews in the future.

Everything looks good to me! I have a separate concern about the colour of the empty-dashed-circle, but I'll create a separate issue since it affects areas outside this element.

**Review by quinnkeast — APPROVED** [review]

**Pull request merged**